language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Code Action for extracting to a constant.
+ *
+ * @since 2201.2.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class ExtractToConstantCodeAction implements RangeBasedCodeActionProvider {
+
+    public static final String NAME = "Extract To Constant";
+    private static final String CONSTANT_NAME_PREFIX = "CONSTANT";
+
+    public List<SyntaxKind> getSyntaxKinds() {
+        return List.of(SyntaxKind.BOOLEAN_LITERAL, SyntaxKind.NUMERIC_LITERAL,
+                SyntaxKind.STRING_LITERAL, SyntaxKind.BINARY_EXPRESSION);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<CodeAction> getCodeActions(CodeActionContext context,
+                                                    RangeBasedPositionDetails posDetails) {
+        
+        if (context.currentSyntaxTree().isEmpty() || context.currentSemanticModel().isEmpty()) {
+            return Collections.emptyList();
+        }
+        Node node = posDetails.matchedCodeActionNode();
+        if (node.parent().kind() == SyntaxKind.CONST_DECLARATION ||
+                node.parent().kind() == SyntaxKind.INVALID_EXPRESSION_STATEMENT ||
+                (node.kind() != SyntaxKind.NUMERIC_LITERAL && node.kind() != SyntaxKind.STRING_LITERAL && 
+                        node.kind() != SyntaxKind.BOOLEAN_LITERAL && node.kind() != SyntaxKind.BINARY_EXPRESSION)) {
+            return Collections.emptyList();
+        }
+        
+        // If the node is a BinaryExpressionNode, check whether it contains only BasicLiteralNodes
+        if (node.kind() == SyntaxKind.BINARY_EXPRESSION) {
+            ConstantVisitor constantVisitor = new ConstantVisitor();
+            node.accept(constantVisitor);
+            if (constantVisitor.getInvalidNode()) {
+                return Collections.emptyList();
+            }
+        }
+        
+        Set<String> visibleSymbolNames = context.visibleSymbols(context.cursorPosition()).stream()
+                .map(Symbol::getName)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toSet());
+        String constName = NameUtil.generateTypeName(CONSTANT_NAME_PREFIX, visibleSymbolNames);
+
+        String value = node.toSourceCode().strip();
+        LineRange replaceRange = node.lineRange();
+        Optional<TypeSymbol> typeSymbol = context.currentSemanticModel().get().typeOf(node);
+        if (typeSymbol.isEmpty()) {
+            return Collections.emptyList();
+        }
+        
+        Node modPartNode = node;
+        while (modPartNode.parent().kind() != SyntaxKind.MODULE_PART) {
+            modPartNode = modPartNode.parent();
+        }
+        
+        String constDeclStr = String.format("const %s %s = %s;%n%n", typeSymbol.get().signature(), constName, value);
+        TextEdit constDeclEdit = new TextEdit(new Range(PositionUtil.toPosition(modPartNode.lineRange().startLine()),
```

Comment:
We cannot add this to the start of the module part node, because imports should come first (otherwise syntax error). We should get either the end position of last import or the start of module part node (if no imports available).

Reply:
Fixed.